src/include/migraphx/op/scatternd_op.hpp

```diff
@@ -28,44 +28,105 @@
+     * is the same shape as the data tensor (first input), but cast to a standard shape.
+     *
+     */
     shape compute_shape(std::vector<shape> inputs) const
     {
-        check_shapes{inputs, *this}.has(3);
-        auto r         = inputs.front().lens().size();
-        auto q         = inputs.at(1).lens().size();
-        auto k         = inputs.at(1).lens().back();
-        auto ind_lens  = inputs.at(1).lens();
-        auto upd_lens  = inputs.back().lens();
-        auto data_lens = inputs.front().lens();
+        check_shapes{inputs, *this, true}.has(3);
+        auto data_shape  = inputs.front();
+        auto index_shape = inputs.at(1);
+        auto upd_shape   = inputs.back();
+
+        auto r = data_shape.ndim();
+        auto q = index_shape.ndim();
+        size_t k;
+        if(index_shape.dynamic())
+        {
+            // the rank of the output is a function of k, so k must be fixed.
+            if(not index_shape.dyn_dims().back().is_fixed())
+            {
+                MIGRAPHX_THROW(
+                    "GATHERND: last dimension of indices tensor must be fixed (min=max)");
+            }
+            k = index_shape.dyn_dims().back().min;
+        }
+        else
+            k = index_shape.lens().back();
+
+        // Checks on the sizes of input tensors
+        if(q + r != upd_shape.ndim() + k + 1)
+            MIGRAPHX_THROW("ScatterND:  ranks of inputs don't match. " + std::to_string(q) + " + " +
+                           std::to_string(r) + " - " + std::to_string(k) +
+                           " - 1 != " + std::to_string(upd_shape.ndim()));
         if(k > r)
             MIGRAPHX_THROW("ScatterND: index of size " + std::to_string(k) +
                            " is too large for tensor of rank " + std::to_string(r));
-        if(not(std::equal(ind_lens.begin(), ind_lens.begin() + q - 1, upd_lens.begin()) and
-               std::equal(data_lens.begin() + k, data_lens.end(), upd_lens.begin() + q - 1)))
-            MIGRAPHX_THROW("ScatterND: incorrect update shape. update.lens != indices.lens[0:q-1] "
-                           "++ data.lens[k:r-1]");
-        auto s = inputs.front();
-        if(s.broadcasted())
+
+        if(migraphx::none_of(inputs, [](auto v) { return v.dynamic(); }))
         {
-            return {s.type(), s.lens()};
+            auto ind_lens  = index_shape.lens();
+            auto upd_lens  = upd_shape.lens();
+            auto data_lens = data_shape.lens();
+
+            // Check that corresponding portions of tensor shapes match
+            if(not(std::equal(ind_lens.begin(), ind_lens.begin() + q - 1, upd_lens.begin()) and
+                   std::equal(data_lens.begin() + k, data_lens.end(), upd_lens.begin() + q - 1)))
+                MIGRAPHX_THROW(
+                    "ScatterND: incorrect update shape. update.lens != indices.lens[0:q-1] "
+                    "++ data.lens[k:r-1]");
```

Comment:
You can do this check only once for both static and dynamic shapes combined.

Reply:
Changed.  This change leads to a conceptual shift that dramatically changed the compute_shape() method.  In the previous code, one section of code deals compares static lengths and the other, very similar, compares dynamic_dimensions.  By calling to_dynamic() on all shapes before comparing, it's possible to do a single comparison that's valid for any combination of types, with much simplified logic.  This might be a good paradigm to use for other operators too.